README.md
# Dynare

Described on the homepage: <http://www.dynare.org/>

Most users should use the precompiled package available for your OS, also
available via the Dynare homepage: <http://www.dynare.org/download/dynare-stable>.

# License

Most of the source files are covered by the GNU General Public Licence version
3 or later (there are some exceptions to this, see [license.txt](license.txt) in
Dynare distribution for specifics).

# Building Dynare From Source

Here, we explain how to build from source:
- Dynare, including preprocessor and MEX files for MATLAB and Octave
- Dynare++
- all the associated documentation (PDF and HTML)

This source can be retrieved in three forms:
- via git, at <https://github.com/DynareTeam/dynare.git>
- using the stable source archive of the latest Dynare version (currently 4.3) from <http://www.dynare.org/download/dynare-4.3/source>
- using a source snapshot of the unstable version, from <http://www.dynare.org/download/dynare-unstable/source-snapshot>

Note that if you obtain the source code via git, you will need to install more tools (see below).

The first section of this page gives general instructions, which apply to all platforms. Then some specific platforms are discussed.

**NB**: Here, when we refer to 32-bit or 64-bit, we refer to the type of MATLAB installation, not the type of Windows installation. It is perfectly possible to run a 32-bit MATLAB on a 64-bit Windows: in that case, instructions for Windows 32-bit should be followed. To determine the type of your MATLAB installation, type:
```matlab
>> computer
```
at the MATLAB prompt: if it returns `PCWIN`, then you have a 32-bit MATLAB; if it returns `PCWIN64`, then you have a 64-bit MATLAB.

**Contents**

1. [**General Instructions**](#general-instructions)
1. [**Debian or Ubuntu**](#debian-or-ubuntu)
1. [**Fedora**](#fedora)
1. [**Windows**](#windows)
1. [**Mac OS X**](#mac-os-x)

## General Instructions

### Prerequisites

A number of tools and libraries are needed in order to recompile everything. You don't necessarily need to install everything, depending on what you want to compile.

- A POSIX compliant shell and an implementation of Make (mandatory)
- The [GNU Compiler Collection](http://gcc.gnu.org/), with gcc, g++ and gfortran (mandatory)
- [MATLAB](http://www.dynare.org/DynareWiki/BuildingDynareFromSource?action=AttachFile&do=view&target=dynare-mingw64-libs.zip) (if you want to compile MEX for MATLAB)
- [GNU Octave](http://www.octave.org), with the development headers (if you want to compile MEX for Octave)
- [Boost libraries](http://www.boost.org), version 1.36 or later
- [Bison](http://www.gnu.org/software/bison/), version 2.3 or later (only if you get the source through Git)
- [Flex](http://flex.sourceforge.net/), version 2.5.4 or later (only if you get the source through Git)
- [Autoconf](http://www.gnu.org/software/autoconf/), version 2.62 or later (only if you get the source through Git) (see [Installing an updated version of Autoconf in your own directory, in GNU/Linux](http://www.dynare.org/DynareWiki/AutoMake))
- [Automake](http://www.gnu.org/software/automake/), version 1.11.2 or later (only if you get the source through Git) (see [Installing an updated version of AutoMake in your own directory, in GNU/Linux](http://www.dynare.org/DynareWiki/AutoMake))
- [CWEB](http://www-cs-faculty.stanford.edu/%7Eknuth/cweb.html), with its tools `ctangle` and `cweave` (only if you want to build Dynare++ and get the source through Git)
- An implementation of BLAS and LAPACK: either [ATLAS](http://math-atlas.sourceforge.net/), [OpenBLAS](http://xianyi.github.com/OpenBLAS/), Netlib ([BLAS](http://www.netlib.org/blas/), [LAPACK](http://www.netlib.org/lapack/)) or [MKL](http://software.intel.com/en-us/intel-mkl/) (only if you want to build Dynare++)
- An implementation of [POSIX Threads](http://en.wikipedia.org/wiki/POSIX_Threads) (optional, for taking advantage of multi-core)
- [MAT File I/O library](http://sourceforge.net/projects/matio/) (if you want to compile Markov-Switching code, the estimation DLL, k-order DLL and Dynare++ in unstable)
- [SLICOT](http://www.slicot.org) (if you want to compile the Kalman steady state DLL)
- [GSL library](http://www.gnu.org/software/gsl/) (if you want to compile Markov-Switching code)
- A decent LaTeX distribution (if you want to compile PDF documentation). The following extra components may be needed:
  - The Econometrica bibliography style: you need [harvard](http://www.ctan.org/tex-archive/macros/latex/contrib/harvard/) and [economic](http://www.ctan.org/tex-archive/biblio/bibtex/contrib/economic/) packages from CTAN (only if you want to build Dynare user guide, no more needed with Dynare unstable)
  - [Eplain](http://www.tug.org/eplain/) TeX macros (only if you want to build Dynare++ source documentation)
  - [Beamer](http://latex-beamer.sourceforge.net/) (for some PDF presentations)
- For building the reference manual:
  - [GNU Texinfo](http://www.gnu.org/software/texinfo/)
  - [Texi2HTML](http://www.nongnu.org/texi2html) and [Latex2HTML](http://www.latex2html.org), if you want nice mathematical formulas in HTML output
  - [Doxygen](http://www.stack.nl/%7Edimitri/doxygen/) (if you want to build Dynare preprocessor source documentation)
- For Octave, the development libraries corresponding to the UMFPACK packaged with Octave (only in unstable)

### Preparing the sources

If you have downloaded the sources from an official source archive or the source snapshot, just unpack it.

If you want to use Git, do the following from a terminal:

    git clone http://github.com/DynareTeam/dynare.git
    cd dynare
    git submodule update --init
    autoreconf -si

The last line runs Autoconf and Automake in order to prepare the build environment (this is not necessary if you got the sources from an official source archive or the source snapshot).

### Configuring the build tree

Simply launch the configure script from a terminal:
```
./configure
```
If you have MATLAB, you need to indicate both the MATLAB location and version. For example, on GNU/Linux:
```
./configure --with-matlab=/usr/local/MATLAB/R2013a MATLAB_VERSION=8.1
```
Note that the MATLAB version can also be specified via the MATLAB family product release (R2009a, R2008b, ...).

**NB**: For MATLAB versions strictly older than 7.1, you need to explicitly give the MEX extension, via `MEXEXT` variable of the configure script (for example, `MEXEXT=dll` for Windows with MATLAB \< 7.1).

Alternatively, you can disable the compilation of MEX files for MATLAB with the `--disable-matlab` flag, and MEX files for Octave with `--disable-octave`.

You may need to specify additional options to the configure script, see the platform specific instructions below.

Note that if you don't want to compile with debugging information, you can specify the `CFLAGS` and `CXXFLAGS` variables to configure, such as:
```
./configure CFLAGS="-O3" CXXFLAGS="-O3"
```
If you want to give a try to the parallelized versions of some mex files (`A_times_B_kronecker_C` and `sparse_hessian_times_B_kronecker_C` used to get the reduced form of the second order approximation of the model) you can add the `--enable-openmp` flag, for instance:
```
./configure --with-matlab=/usr/local/matlab78 MATLAB_VERSION=7.8 --enable-openmp
```
If the configuration goes well, the script will tell you which components are correctly configured and will be built.

### Bulding

Binaries and Info documentation are built with:
```
make
```
PDF and HTML documentation are respectively built with:
```
make pdf
make html
```
The testsuites can be run with:
```
make check
```
## Debian or Ubuntu

All the prerequisites are packaged.

The easiest way to install the pre-requisites in Debian is to use Debian's dynare package and do:
```
apt-get build-dep dynare
```
Alternatively, if you want to build everything, manually install the following packages:

- `build-essential` (for gcc, g++ and make)
- `octave3.2-headers` or `liboctave-dev` (will install ATLAS)
- `libboost-graph-dev`
- `libgsl0-dev`
- `libmatio-dev`
- `libslicot-dev` and `libslicot-pic`
- `libsuitesparse-dev` (only for Unstable)
- `flex`
- `bison`
- `autoconf`
- `automake`
- `texlive`
- `texlive-publishers` (for Econometrica bibliographic style)
- `texlive-extra-utils` (for CWEB)
- `texlive-formats-extra` (for Eplain)
- `texlive-latex-extra` (for fullpage.sty)
- `latex-beamer`
- `texinfo`
- `texi2html`, `latex2html`
- `doxygen`

## Fedora

**NB**: Documentation still in progress…
- `octave-devel`
- `boost-devel`
- `gsl-devel`
- `matio-devel`
- `flex`
- `bison`
- `autoconf`
- `automake`
- `texlive`
- `texinfo`
- `texi2html`, `latex2html`
- `doxygen`

## Windows

The following instructions are compatible with MATLAB or with Octave/MinGW (as downloadable [here](http://www.dynare.org/download/octave)).

### Setting up the Compilation Environment

- First, you need to setup a Cygwin environment, following the instructions at <http://www.cygwin.com>. You need the following packages:
    - `make`
    - `bison`
    - `flex`
    - `autoconf` and `autoconf2.5`
    - `automake` and `automake1.11`
    - `texlive`, `texlive-collection-latexextra`, `texlive-collection-formatsextra`, `texlive-collection-publishers`
    - `texinfo`
    - `doxygen`
    - `mingw64-i686-gcc`, `mingw64-i686-gcc-g++`, `mingw64-i686-gcc-fortran` (if you have Octave/MinGW or if you have MATLAB 32-bit)
    - `mingw64-x86_64-gcc`, `mingw64-x86_64-gcc-g++`, `mingw64-x86_64-gcc-fortran` (if you have MATLAB 64-bit)
- Second, install precompiled librairies for BLAS, LAPACK, Boost and GSL:
    - If you have Octave or MATLAB 32-bit, download [dynare-mingw32-libs.zip](http://www.dynare.org/DynareWiki/BuildingDynareFromSource?action=AttachFile&do=view&target=dynare-mingw32-libs.zip), and uncompress it in `c:\cygwin\usr\local\lib\mingw32`
    - If you have MATLAB 64-bit, download [dynare-mingw64-libs.zip](http://www.dynare.org/DynareWiki/BuildingDynareFromSource?action=AttachFile&do=view&target=dynare-mingw64-libs.zip), and uncompress it in `c:\cygwin\usr\local\lib\mingw64`

### Compiling the preprocessor, Dynare++, the MEX for MATLAB and the documentation

Download and uncompress the Dynare source tree, let’s say in `c:\cygwin\home\user\dynare`.

Launch a Cygwin shell, and enter the Dynare source tree:
```
cd dynare
```
If you retrieved the source from Git, don't forget to do:
```
autoreconf -i -s
```
Then, configure the package.

- If your MATLAB is 32-bit, let's say version R2008b installed in `c:\Program Files\MATLAB\R2008b`
```
./configure --host=i686-w64-mingw32 --with-boost=/usr/local/lib/mingw32/boost --with-blas=/usr/local/lib/mingw32/blas/libopenblas.a --with-lapack=/usr/local/lib/mingw32/lapack/liblapack.a --with-gsl=/usr/local/lib/mingw32/gsl --with-matio=/usr/local/lib/mingw32/matio --with-slicot=/usr/local/lib/mingw32/slicot --with-matlab=/cygdrive/c/Progra~1/MATLAB/R2008b MATLAB_VERSION=R2008b --disable-octave
```
- If your MATLAB is 64-bit:
```
./configure --host=x86_64-w64-mingw32 --with-boost=/usr/local/lib/mingw64/boost --with-blas=/usr/local/lib/mingw64/blas/libopenblas.a --with-lapack=/usr/local/lib/mingw64/lapack/liblapack.a --with-gsl=/usr/local/lib/mingw64/gsl --with-matio=/usr/local/lib/mingw64/matio --with-slicot=/usr/local/lib/mingw64/slicot --with-matlab=/cygdrive/c/Progra~1/MATLAB/R2008b MATLAB_VERSION=R2008b --disable-octave
```
A few remarks:

- Note that here we use `Progra~1` (the 8.3 filename) instead of `Program Files`. This is because spaces in filenames confuse the configuration scripts.
- If you don’t have MATLAB, then drop the `--with-matlab` and `MATLAB_VERSION` options
- If your MATLAB is 32-bit and your Windows is 64-bit, you need to explicitly give the MEX extension, with `MEXEXT=mexw32`

Then compile everything with:
```
make all pdf html
```
This should build:

- Dynare preprocessor
- Dynare MEX files for MATLAB (provided you gave the MATLAB path to configure)
- Dynare++
- Part of the documentation

### Compiling the MEX for Octave (MinGW package)

Launch a Cygwin shell, and enter the Dynare source tree for Octave MEX:

    cd dynare/mex/build/octave

Configure and make:

    ./configure MKOCTFILE=/usr/local/lib/mingw32/mkoctfile-win --with-boost=/usr/local/lib/mingw32/boost --with-gsl=/usr/local/lib/mingw32/gsl --with-matio=/usr/local/lib/mingw32/matio --with-slicot=/usr/local/lib/mingw32/slicot-underscore
    make

## Mac OS X

- Install the Xcode Common Tools:
    - Install [Xcode](http://developer.apple.com/xcode/) from the App Store
    - Open Xcode
    - Go to `Xcode->Preferences...`
    - In the window that opens, click on the `Downloads` tab
    - In the tab that appears, click on the `Components` button
    - Next to `Command Line Tools`, click on `Install`
- Download [MacOSX10.6.sdk.zip](http://www.jamesgeorge.org/uploads/MacOSX10.6.sdk.zip) and unzip it in `/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs`. Change the owner to be `root` and the group to be `wheel`

- Install [Homebrew](http://mxcl.github.io/homebrew/) and [Homebrew Science](https://github.com/Homebrew/homebrew-science)
- Install the following brews:
```
    brew install automake
    brew install gsl
    brew install boost
    brew install gfortran
    brew install matlab2tikz --HEAD
    brew install libmatio --with-hdf5
    brew install slicot --with-default-integer-8
```
- **(Optional)** To compile Dynare mex files for use on Octave, first install Octave following the [Simple Installation Instructions](http://wiki.octave.org/Octave_for_MacOS_X#Simple_Installation_Instructions_3). Then, you will probably also want to install graphicsmagick via Homebrew with `brew install graphicsmagick`.
- **(Optional)** To compile Dynare's documentation, first install the latest version of [MacTeX](http://www.tug.org/mactex/). Then install `doxygen` and `latex2html` via Homebrew with the following commands:
- **(On OS X 10.7 Only)** Copy [FlexLexer.h](http://www.dynare.org/DynareWiki/BuildingDynareFromSource?action=AttachFile&do=view&target=FlexLexer.h) into the `preprocessor` directory (there was an error in the `FlexLexer.h` file distributed with 10.7)
- Finally, switch to the root dynare directory. Ensure your path contains `/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/bin:/opt/X11/bin:/usr/texbin:/usr/local/sbin`. Run:
    - `autoconf -si`
    - `./configure --with-matlab=/Applications/MATLAB_R2013a.app MATLAB_VERSION=8.1` for builds with Matlab or `./configure` for builds just using Octave
    - `make`


debian/changelog
dynare (4.0.4) unstable; urgency=low

  * Fixed forecasting after stoch_simul
  * Fixed crash of the smoother when only one exogenous variable
  * Fixed issue with compilation of MEX files under 64 bits platforms with Matlab R2009a
  * Fixed name clash when MOD file is called "ramsey.mod" and computes optimal policy
  * Fixed USE_DLL option at order 2
  * Updates to the reference manual in sections on model resolution and estimation

 -- Sébastien Villemot <sebastien.villemot@ens.fr>  Fri, 19 Jun 2009 16:48:54 +0200

dynare (4.0.3) unstable; urgency=low

  * Many bugfixes, including:
    - fixed estimation of constants in linear models
    - determistic exogenous now work when there is more than one such variable
    - fixed recursive estimation
    - fixed OSR
    - fix to interface to Global Sensity Analysis toolbox
    - "simul" and "periods" options are now accepted simultaneously by "stoch_simul"
    - preprocessor no longer crashes when there is an unexpected EOF
    - operator precedence is now correct in macro-processor
    - estimated_params_init and estimated_params_bounds now work in conjunction with correlation of endogenous variables
    - fixed bug in the macro-processor, when the body of a @#for loop is empty
    - fixed memory management bug in the preprocessor related to the filename being parsed
    - fixing bug for mode_compute=5,6
  * build_matlab.m script works on Matlab for Windows 64 bits
  * Structures M_ and options_ are saved in the "_results.mat" output file
  * Structure bayestopt_ is saved in the "_mode.mat" output file
  * Added option "notmpterms" to preprocessor, for disabling computation of temporary terms
  * Macro processor reports the location of @#for or @#if statement when not matched by an @#endfor or an @#endif
  * Many updates to the reference manual for 4.0 (which is not yet finished)
  * All code source is now under a free license (GNU General Public License 3 for most of them)
  * All documentation (including user guide) now licensed under the GNU Free Documentation License 1.3

 -- Sébastien Villemot <sebastien.villemot@ens.fr>  Tue, 24 Mar 2009 10:48:36 +0100

dynare (4.0.2) unstable; urgency=low

  * Fixed bugs in Kalman filter and smoother
  * Fixed bug with "corr" statement of "shocks" block
  * Added error message if number of equations differs from number of endogenous variables (except when doing optimal policy computation, or standalone BVAR estimation)
  * Fixed bugs with "dynatype" and "dynasave" commands
  * Fixed name collisions with Matlab functions "sec2hms" and "bicgstab"
  * Fixed some licensing issues
  * Removed obsolete "dr_algo" option
  * Fixed crashes under Octave due to graphics limitations
  * Added error message when model local variables (pound sign expressions) are used outside model scope
  * Included AIM subdirectory in packaging
  * Estimation: modified computation of log(det(invhess)) to solve cases in large models where det(invhess) is smaller than machine precision

 -- Sébastien Villemot <sebastien.villemot@ens.fr>  Tue, 28 Oct 2008 14:57:28 +0100

dynare (4.0.1) unstable; urgency=low

  * Include matlab/distributions/ directory in packaging
  * BVAR à la Sims: the preprocessor now accepts integers for tau and
    lambda parameters (thanks to Aqua for reporting)
  * More information messages displayed during homotopy (modes 1,2,3)
  * Fixed bugs in homotopy mode 3
  * Fixed typo in DiffuseKalmanSmootherH1.m (thanks to Tsu-ting Tim Lin
    for reporting)
  * Fixed bug with cumulative normal distribution in model block (thanks
    to Karl Walentin for reporting)
  * Changes to Dynare solver:
    - more debugging messages with solve_algo=2
    - solve_algo=2 now exits immediately after failure to solve one block
    - new solve_algo=4, similar to solve_algo=2 but with a different way
      of dealing with badly scaled or nearly singular Jacobian
    - solve_algo=0 fails if Matlab's Optimization Toolbox not present
    - solver fails if solve_algo not in [0,1,2,3,4]
	
 -- Sébastien Villemot <sebastien.villemot@ens.fr>  Tue, 23 Sep 2008 11:41:41 +0200

dynare (4.0.0) UNRELEASED; urgency=low

  * First official release in the 4.0 series
  * New functions {load,save}_params_and_steady_state
  * Fixed bug with forecasting with exogenous deterministic variables
  * BVAR à la Sims:
    - no longer remove explosive models when drawing from the posterior
    - forecasting now reports median path instead of mean path
  * Fixed bug and improve performance in diffuse filter/smoother
  * Fixed crash of estimation in some cases
	
 -- Sébastien Villemot <sebastien.villemot@ens.fr>  Mon, 15 Sep 2008 11:15:20 +0200

dynare (4.0~rc1) UNRELEASED; urgency=low

  * First release candidate for 4.0 series
  * Updates to the manual:
    - added empty entries for commands new in Dynare 4
    - removed obsolete commands
    - improved description of valid mathematical expressions
  * Added interface to Gary Anderson's AIM package (not yet documented)
  * Fix copyright/license headers for most source files
  * Replaced functions related to probability distributions (CDF, PDF, quantile)
    by GPL'd ones
  * Added preprocessor support for inverse gamma of type 2
  * New command "model_info" providing informations about the block structure
    of the model
  * Informations about determistic simulations (number of iterations, error...)
    are now saved in oo_
  * Added Independent Metropolis Hastings and multivariate student as an
    alternative to multivariate normal for the proposal distribution. The
    default posterior sampling algorithm is random walk metropolis with a
    gaussian proposal distribution.
  * Added trace plots
  * Added autocorrelogram for the posterior draws
  * If no variable given to "estimation" command, the user is asked to confirm
    that she wants to obtain posterior statistics for all the endogenous
    variables of the model. The user can then choose between:
    - the set of all the endogenous variables

matlab/osr.m
% Copyright (C) 2001 Michel Juillard
%
function osr(var_list,params,W)
  global iter_ ys_ dr_ y_ dr_ exo_nbr lgy_ lgx_ Sigma_e_ ykmin_ ykmax_ ...
      endo_nbr options_ lgx_orig_ord_
  
  options_old = options_;
  options_.order = 1;
  options_.linear = 1;
  options_ = set_default_option(options_,'ar',5);
  options_ = set_default_option(options_,'irf',40);
  options_ = set_default_option(options_,'dr_algo',0);
  options_ = set_default_option(options_,'simul_algo',0);
  options_ = set_default_option(options_,'drop',100);
  options_ = set_default_option(options_,'replic',1);
  options_ = set_default_option(options_,'nomoments',0);
  options_ = set_default_option(options_,'nocorr',0);
  options_ = set_default_option(options_,'simul_seed',0);
  options_ = set_default_option(options_,'hp_filter',0);
  options_ = set_default_option(options_,'hp_ngrid',512);
  options_ = set_default_option(options_,'simul',0);
  options_ = set_default_option(options_,'periods',1);
  options_ = set_default_option(options_,'TeX',0);
  options_ = set_default_option(options_,'noprint',0);
  if options_.simul & ~isempty(iter_) & options_.periods == 0
    options_.periods = iter_;
  end

  iter_ = max(options_.periods,1);
  options_.periods = iter_;
  
  make_ex_;

  disp(' ')
  disp('OPTIMAL SIMPLE RULE')
  disp(' ')
  [dr_,info] = osr1(params,W);
  if info > 0
    options_ = options_old;
    return
  elseif options_.noprint ~= 1
      disp('MODEL SUMMARY')
      disp(' ')
      disp(['  Number of variables:         ' int2str(endo_nbr)])
      disp(['  Number of stochastic shocks: ' int2str(exo_nbr)])
      disp(['  Number of state variables:   ' ...
	    int2str(length(find(dr_.kstate(:,2) <= ykmin_+1)))])
      disp(['  Number of jumpers:           ' ...
	    int2str(length(find(dr_.kstate(:,2) == ykmin_+2)))])
      disp(['  Number of static variables:  ' int2str(dr_.nstatic)])
      my_title='MATRIX OF COVARIANCE OF EXOGENOUS SHOCKS';
      labels = deblank(lgx_);
      headers = strvcat('Variables',labels);
      lh = size(labels,2)+2;
      table(my_title,headers,labels,Sigma_e_,lh,10,6);
      disp(' ')
      disp_dr(dr_,options_.order,var_list);
      if options_.order == 1 & options_.simul == 0 & options_.nomoments == 0
	disp_th_moments(dr_,var_list);
      elseif options_.simul == 1 | options_.nomoments == 0
	if options_.periods == 0
	  error('OSR error: number of periods for the simulation isn''t specified')
	end
	if options_.periods < options_.drop
	  disp(['OSR error: The horizon of simulation is shorter' ...
		' than the number of observations to be DROPed'])
	  return
	end
	
	y_ = simult(repmat(dr_.ys,1,ykmin_),dr_);
	dyn2vec;
	if options_.nomoments == 0
	  disp_moments(y_,var_list);
	end
      end
      
      n = size(var_list,1);
      if n == 0
	n = length(dr_.order_var);
	ivar = [1:n]';
	var_list = lgy_;
      else
	ivar=zeros(n,1);
	for i=1:n
	  i_tmp = strmatch(var_list(i,:),lgy_,'exact');
	  if isempty(i_tmp)
	    error (['One of the specified variables does not exist']) ;
	  else
	    ivar(i) = i_tmp;
	  end
	end
      end

      if n < 13 & options_.irf > 0
	if n == 1
	  nr = 1;
	  nc = 1;
	elseif n == 2
	  nr = 1;
	  nc = 2;
	elseif n <= 4
	  nr = 2;
	  nc = 2;
	elseif n <= 6
	  nr = 2;
	  nc = 3;
	elseif n <= 9
	  nr = 3;
	  nc = 3;
	elseif n <= 12
	  nr = 3;
	  nc = 4;
	end
	olditer = iter_;
	if options_.order == 1
	  options_.replic = 1;
	else
	  if options_.replic == 0
	    options_.replic = 50;
	  end
	end
	SS(lgx_orig_ord_,lgx_orig_ord_)=Sigma_e_+1e-14*eye(exo_nbr);
	cs = transpose(chol(SS));

	for i = 1:exo_nbr
	  if SS(i,i) > 1e-13
	    figure('Name',['Shock to ' lgx_(i,:)]);
	    y=irf(dr_,cs(lgx_orig_ord_,i), options_.irf, options_.drop, options_.replic, options_.order);
	    for j = 1:n
	      subplot(nr,nc,j);
	      plot([y(ivar(j),:)']);
	      title(var_list(j,:),'Interpreter','none');
	      assignin('base',[deblank(var_list(j,:)) '_' deblank(lgx_(i,:))],y(ivar(j),:)');
	    end
	  end
	  
	end
	iter_ = olditer;
      end
  end
  options_ = options_old;
  
% 01/10/01 FC dr_ and y_ made global
% 02/20/01 MJ ys_ removed from calling sequence for simult (all in dr_)
% 02/23/01 MJ added dyn2vec()
% 06/24/01 MJ steady -> steady_
% 09/24/01 MJ dr_ made global
% 08/28/02 MJ added var_list
% 10/09/02 MJ no simulation and theoretical moments for order 1 
% 10/14/02 MJ added plot of IRFs
% 10/30/02 MJ options_ are now a structure
% 01/01/03 MJ added dr_algo
% 01/09/03 MJ set default values for options_ (correct absence of autocorr
%             when order == 1)
% 01/12/03 MJ removed call to steady_ as already checked in resol()
% 01/31/03 MJ make IRF global with varname_shockname
% 02/09/03 MJ ys_ reset with value declared in initval after computations
% 02/18/03 MJ removed above change. ys_ shouldn't be affected by
%             computations in this function
%             new option SIMUL computes a stochastic simulation and save
%             results in y_ and via dyn2vec
% 04/03/03 MJ corrected bug for simulation with ykmin_ > 1
